akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -136,51 +208,79 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
               bufferedFrameOut.tryFlush()
 
             case e: StreamFrameEvent if !Http2Compliance.isClientInitiatedStreamId(e.streamId) ⇒
-              pushGOAWAY()
+              pushGOAWAY(ErrorCode.PROTOCOL_ERROR)
 
             case e: StreamFrameEvent if e.streamId > closedAfter.getOrElse(Int.MaxValue) ⇒
             // streams that will have a greater stream id than the one we sent with GO_AWAY will be ignored
 
             case headers @ HeadersFrame(streamId, endStream, endHeaders, fragment) if lastStreamId < streamId ⇒
               val subSource = new SubSourceOutlet[StreamFrameEvent](s"substream-out-$streamId")
-              val handler = SubStream(streamId, headers, StreamState.Open /* FIXME stream state */ , subSource, None, streamLevelWindow)
+              // A HEADERS frame carries the END_STREAM flag that signals the end of a stream.
+              // However, a HEADERS frame with the END_STREAM flag set can be followed by CONTINUATION
+              // frames on the same stream. Logically, the CONTINUATION frames are part of the HEADERS frame.
+              val state = if (endStream && endHeaders) StreamState.HalfClosedRemote else StreamState.Open
+              val handler = SubStream(streamId, headers, state, subSource, None, streamLevelWindow)
+              if (!endHeaders) {
+                handler.continuation = true
+              }
               incomingStreams += streamId → handler // TODO optimise for lookup later on
 
               val sub = makeHttp2SubStream(handler)
               if (sub.initialFrame.endHeaders) dispatchSubstream(sub)
             // else we're expecting some continuation frames before we kick off the dispatch
-            // FIXME: handle the correct order of frames
-
-            case h: HeadersFrame ⇒
-              pushGOAWAY()
 
             case e: StreamFrameEvent if !incomingStreams.contains(e.streamId) ⇒
               // if a stream is invalid we will GO_AWAY
-              pushGOAWAY()
+              pushGOAWAY(ErrorCode.PROTOCOL_ERROR)
+
+            case h: HeadersFrame if isHalfClosedOrClosed(h.streamId) ⇒
+              // stream is half-closed or closed
+              pushGOAWAY(ErrorCode.STREAM_CLOSED)
+
+            case h: HeadersFrame ⇒
+              // stream is open
+              pushGOAWAY(ErrorCode.STREAM_CLOSED)
```

Comment:
Guess it's also easier to remove isHalfClosedOrClosed and switch it with isOpen....